src/main/java/org/hibernate/models/internal/dynamic/DynamicAnnotationUsage.java

```diff
@@ -98,15 +110,76 @@ public <V> V setAttributeValue(String name, V value) {
 			values = new HashMap<>();
 		}
 
+		if ( value != null ) {
+
+			//noinspection unchecked
+			return (V) values.put( name, value );
+		}
 		//noinspection unchecked
-		return (V) values.put( name, value );
+		return (V) values;
 	}
 
-	private static <A extends Annotation> Map<String, Object> extractBaselineValues(AnnotationDescriptor<A> annotationDescriptor) {
+	private static <A extends Annotation> Map<String, Object> extractBaselineValues(
+			AnnotationDescriptor<A> annotationDescriptor,
+			AnnotationTarget target,
+			SourceModelBuildingContext context) {
 		final HashMap<String, Object> values = new HashMap<>();
 		for ( AttributeDescriptor<?> attribute : annotationDescriptor.getAttributes() ) {
-			values.put( attribute.getName(), attribute.getAttributeMethod().getDefaultValue() );
+			values.put( attribute.getName(), getDefaultValue( attribute, target, context ) );
 		}
 		return values;
 	}
+
+	private static Object getDefaultValue(
+			AttributeDescriptor<?> attribute,
+			AnnotationTarget target,
+			SourceModelBuildingContext context) {
+		final Object defaultValue = attribute.getAttributeMethod().getDefaultValue();
+		Object annotation = getValue( defaultValue, target, context );
+		if ( annotation != null ) {
+			return annotation;
+		}
+		return defaultValue;
+	}
+
+	private static Object getValue(Object value, AnnotationTarget target, SourceModelBuildingContext context) {
```

Comment:
Not a fan of the method name, though not immediately sure of a better one.  `#wrapValue`?

Reply:
changing it and fixing a couple of other issues